sklearn/datasets/_base.py

```diff
@@ -203,7 +237,11 @@ def load_files(
     for label, folder in enumerate(folders):
         target_names.append(folder)
         folder_path = join(container_path, folder)
-        documents = [join(folder_path, d) for d in sorted(listdir(folder_path))]
+        documents = [
+            join(folder_path, d)
+            for d in sorted(listdir(folder_path))
+            if _check_valid_document(d, allowed_extensions)
+        ]
```

Comment:
With `_check_valid_document` being so small, I think we can split on `allowed_extensions` here:

```python
files = sorted(listdir(folder_path))
if allowed_extensions is not None:
    allowed_extensions = frozenset(allowed_extensions)
    documents = [join(...) for file in files if os.path.splitext(file)[1] in allowed_extensions]
else:
    documents = [join(...) for file in files]
    
```